### PR TITLE
Add inclined no displacement bc ACTION

### PR DESCRIPTION
**Pull request by jiangwen84**
#13128

### CODE REVIEW

File: modules/tensor_mechanics/src/actions/InclinedNoDisplacementBCAction.C [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  params.addRequiredParam<Real>("penalty", "Penalty parameter");
+  return params;
+}
+
+InclinedNoDisplacementBCAction::InclinedNoDisplacementBCAction(const InputParameters & params)
+  : Action(params)
+{
+  _save_in_vars.push_back(getParam<std::vector<AuxVariableName>>("save_in_disp_x"));
+  _save_in_vars.push_back(getParam<std::vector<AuxVariableName>>("save_in_disp_y"));
+  _save_in_vars.push_back(getParam<std::vector<AuxVariableName>>("save_in_disp_z"));
```
Comment (permcody) [inline modules/tensor_mechanics/src/actions/InclinedNoDisplacementBCAction.C]: Without a default, doesn't this error when it's not provided?

File: modules/tensor_mechanics/src/actions/InclinedNoDisplacementBCAction.C [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  const std::string kernel_name = "PenaltyInclinedNoDisplacementBC";
+
+  std::vector<VariableName> displacements;
+  if (isParamValid("displacements"))
+    displacements = getParam<std::vector<VariableName>>("displacements");
+  else
+  {
+    // Legacy parameter scheme for displacements
+    if (!isParamValid("disp_x"))
+      mooseError("Specify displacement variables using the `displacements` parameter.");
```
Comment (permcody) [inline modules/tensor_mechanics/src/actions/InclinedNoDisplacementBCAction.C]: mooseError -> paramError

File: modules/tensor_mechanics/src/actions/InclinedNoDisplacementBCAction.C [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      if (isParamValid("disp_z"))
+        displacements.push_back(getParam<VariableName>("disp_z"));
+    }
+  }
+
+  if (displacements.size() == 1)
+    mooseError("InclinedNoDisplacementBC is specific to 2D and 3D models.");
+
+  // Create pressure BCs
+  for (unsigned int i = 0; i < displacements.size(); ++i)
```
Comment (permcody) [inline modules/tensor_mechanics/src/actions/InclinedNoDisplacementBCAction.C]: Prefer using MooseIndex now or at least the right type :wink:

File: modules/tensor_mechanics/test/tests/inclined_bc/tests [hunk 4]
```diff
@@ -4,15 +4,24 @@
...
+    requirement = 'The TensorMechanics module shall have the capabilty to enforce inclined boundary conditions on a 3D model using a penalty method.'
+  [../]
+  [./3D_action]
+    type = Exodiff
+    input = 'inclined_bc_action.i'
+    exodiff = 'inclined_bc_3d_out.e'
+    prereq = '3D'
+    issues = '#13128'
+    design = 'InclinedNoDisplacementBCAction.md'
+    requirement = 'The InclinedNoDisplacement boundary condition action shall create the objects needed to apply inclined boundary conditions on a 3D model using a penalty method.'
```
Comment (permcody) [inline modules/tensor_mechanics/test/tests/inclined_bc/tests]: This test requires superLU - make sure you add superlu = true in this test spec.

File: modules/tensor_mechanics/src/actions/InclinedNoDisplacementBCAction.C (hunk 1 above)
Comment (jiangwen84) [reply modules/tensor_mechanics/src/actions/InclinedNoDisplacementBCAction.C]: I tested it and it does not error :-)

_save_in_vars[i] will be empty if it is not provided.

File: modules/tensor_mechanics/doc/content/source/actions/InclinedNoDisplacementBCAction.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+# InclinedNoDisplacementBC Action
+
+## Description
+
+The `InclinedNoDisplacementBC` Action is used to create a set of pressure boundary conditions for a string of displacement variables. See the description, example use, and parameters on the [InclinedNoDisplacementBC](/InclinedNoDisplacementBC/index.md) action system page.
```
Comment (bwspenc) [inline modules/tensor_mechanics/doc/content/source/actions/InclinedNoDisplacementBCAction.md]: pressure -> InclinedNoDisplacementBC

File: modules/tensor_mechanics/doc/content/syntax/BCs/InclinedNoDisplacementBC/index.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
+# Inclined No Displacement Boundary Condition Action System
+
+!syntax description /BCs/InclinedNoDisplacementBC/InclinedNoDisplacementBCAction
+
+## Description
+
+The InclinedNoDisplacementBCAction Action, given in the input file as simply `InclinedNoDisplacementBC`, is designed to simplify the input file when several variables have the same inclined no displacement boundary condition [Inclined no displacement boundary condition](PenaltyInclinedNoDisplacementBC.md)  applied in the normal component.
+
+The Pressure Action, given in the input file as simply `Pressure`, is designed to simplify the input file when several variables have the same pressure boundary condition magnitude applied in the normal component.
```
Comment (bwspenc) [inline modules/tensor_mechanics/doc/content/syntax/BCs/InclinedNoDisplacementBC/index.md]: I think you want to remove these two lines.

File: modules/tensor_mechanics/src/actions/InclinedNoDisplacementBCAction.C [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+InputParameters
+validParams<InclinedNoDisplacementBCAction>()
+{
+  InputParameters params = validParams<Action>();
+  params.addClassDescription("Set up inclined no displacement boundary conditions");
+
+  params.addRequiredParam<std::vector<BoundaryName>>(
+      "boundary", "The list of boundary IDs from the mesh where the pressure will be applied");
+
+  params.addParam<VariableName>("disp_x", "The x displacement");
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/actions/InclinedNoDisplacementBCAction.C]: Get rid of the independent `disp_x`, `disp_y`, and `disp_z` parameters. That's the old way of doing this, and we're getting rid of it.

File: modules/tensor_mechanics/src/actions/InclinedNoDisplacementBCAction.C [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+  params.addParam<VariableName>("disp_y", "The y displacement");
+  params.addParam<VariableName>("disp_z", "The z displacement");
+
+  params.addParam<std::vector<VariableName>>(
+      "displacements",
+      "The displacements appropriate for the simulation geometry and coordinate system");
+
+  params.addParam<std::vector<AuxVariableName>>("save_in_disp_x",
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/actions/InclinedNoDisplacementBCAction.C]: Same with this.  Handle the save_in variables in the same way that it is done in TensorMechanicsAction.C (single input parameter `save_in` takes a vector of variables).  It looks like the Pressure action (which I assume you based this off of) is not set up that way, but it should be.

File: modules/tensor_mechanics/test/tests/inclined_bc/inclined_bc_2d.i [hunk 9]
```diff
@@ -35,28 +35,28 @@
     [../]
   [../]
   [./right_x]
-    type = PenaltyInclinedBC
+    type = PenaltyInclinedNoDisplacementBC
```
Comment (bwspenc) [inline modules/tensor_mechanics/test/tests/inclined_bc/inclined_bc_2d.i]: I don't think it's really worth the trouble of having a non-Action version of these tests. Just have a 2D and a 3D test that use the Action.

**Review by bwspenc — CHANGES_REQUESTED** [review]

File: modules/tensor_mechanics/test/tests/inclined_bc/inclined_bc_2d.i (hunk 9 above)
Comment (jiangwen84) [reply modules/tensor_mechanics/test/tests/inclined_bc/inclined_bc_2d.i]: done

File: modules/tensor_mechanics/src/actions/InclinedNoDisplacementBCAction.C (hunk 8 above)
Comment (jiangwen84) [reply modules/tensor_mechanics/src/actions/InclinedNoDisplacementBCAction.C]: done!

File: modules/tensor_mechanics/src/actions/InclinedNoDisplacementBCAction.C (hunk 7 above)
Comment (jiangwen84) [reply modules/tensor_mechanics/src/actions/InclinedNoDisplacementBCAction.C]: done

File: modules/tensor_mechanics/doc/content/syntax/BCs/InclinedNoDisplacementBC/index.md (hunk 6 above)
Comment (jiangwen84) [reply modules/tensor_mechanics/doc/content/syntax/BCs/InclinedNoDisplacementBC/index.md]: done

**Review by bwspenc — APPROVED** [review]

**Pull request merged**